syntax = "proto3";
import "google/protobuf/empty.proto";

package chat;

service UserService {
  // Methods to register, delete and find users
  rpc RegisterUser(RegisterUserMessageRequest) returns (RegisterMessageReply) {}
  rpc RemoveUser(LookupUserRequest) returns (google.protobuf.Empty) {}
  rpc LookupUser(LookupUserRequest) returns (LookupUserReply) {}

  // Methods to register and find group chats
  rpc RegisterGroup(RegisterGroupMessageRequest) returns (RegisterMessageReply) {}
  rpc LookupGroup(LookupGroupRequest) returns (LookupGroupReply) {}

  // Methods to manage users connected to a group
  rpc AddUserToGroup(LookupGroupRequest) returns (LookupGroupReply) {}
  rpc DeleteUserFromGroup(LookupGroupRequest) returns (LookupGroupReply) {}
}

// The request message for registering a user.
message RegisterUserMessageRequest {
    string username = 1;
    string ip = 2;
    int64 port = 3;
}

// The reply message containing the registration status.
message RegisterMessageReply {
  string message = 1;
}

// The request message for looking up a user.
message LookupUserRequest {
  string username = 1;
}

// The reply message containing the lookup result.
message LookupUserReply {
  bool status = 1;
  string username = 2;
  string ip = 3;
  int64 port = 4;
}

// The request message for registering a group.
message RegisterGroupMessageRequest {
    string group_name = 1;
}

// The request message for looking up a user.
message LookupGroupRequest {
  string group_name = 1;
}

// The reply message containing the lookup result.
message LookupGroupReply {
  bool status = 1;
  string group_name = 2;
  int64 num_users = 3;
}